moler/cmd/unix/iptables.py

```diff
@@ -65,32 +68,59 @@ def _parse_chain_references(self, line):
             self.current_ret[self.chain]["REFERENCES"] = self._regex_helper.group("REFERENCES")
             self.current_ret[self.chain]["CHAIN"] = []
+            self._key_details = list()
+            raise ParsingDone
+
+    # Chain INPUT (policy ACCEPT)
+    _re_parse_chain_policy = re.compile(
+        r"Chain\s+(?P<NAME>\S+)\s+\(policy\s(?P<POLICY>\S+)\)$")
+
+    def _parse_chain_policy(self, line):
+        if self._regex_helper.search_compiled(Iptables._re_parse_chain_policy, line):
+            self.chain = self._regex_helper.group("NAME")
+            self.current_ret[self.chain] = dict()
+            self.current_ret[self.chain]["POLICY"] = self._regex_helper.group("POLICY")
+            self.current_ret[self.chain]["CHAIN"] = []
+            self._key_details = list()
             raise ParsingDone
 
     #    pkts      bytes target     prot opt in     out     source               destination
     _re_parse_headers = re.compile(r"(?P<HEADERS>pkts\s+bytes\s+target\s+prot\s+opt\s+in\s+out\s+source\s+destination)")
 
     def _parse_headers(self, line):
-        if self._regex_helper.search_compiled(Iptables._re_parse_headers, line):
+        if self.chain and 0 == len(self._key_details):
+            matched = re.findall(r"\s*(\S+)\s*", line)
+            if matched:
+                for header in matched:
+                    self._key_details.append(header.upper())
             raise ParsingDone
+        # if self._regex_helper.search_compiled(Iptables._re_parse_headers, line):
+        #    raise ParsingDone
 
     #   0        0 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0            icmptype 8 limit: avg 25/sec burst 5
-    _re_parse_details = re.compile(r"\s+(?P<VALUE>\S+)")
-    _re_parse_rest = re.compile(r"(?P<DETAIL>\s+\S+){9}\s+(?P<REST>.*)")
-    _key_details = ["PKTS", "BYTES", "TARGET", "PROT", "OPT", "IN", "OUT", "SOURCE", "DESTINATION"]
+    _re_parse_details = re.compile(r"(?P<VALUE>\S+)")
 
     def _parse_details(self, line):
-        if self._regex_helper.search_compiled(Iptables._re_parse_details, line):
+        if self.chain and self._key_details and self._regex_helper.search_compiled(Iptables._re_parse_details, line):
             values = re.findall(Iptables._re_parse_details, line)
             ret = dict()
-            for value, key in zip(values, Iptables._key_details):
-                ret[key] = value
-            self.current_ret[self.chain]["CHAIN"].append(ret)
-        if self._regex_helper.search_compiled(Iptables._re_parse_rest, line):
-            ret = dict()
-            ret["REST"] = self._regex_helper.group("REST")
-            self.current_ret[self.chain]["CHAIN"].append(ret)
-            raise ParsingDone
+            if len(values) >= len(self._key_details):
+                for value, key in zip(values, self._key_details):
+                    ret[key] = value
+                self.current_ret[self.chain]["CHAIN"].append(ret)
+                regex_for_rest = ""
```

Comment:
Could you extraxt below code to private method?
```
                regex_for_rest = ""
                i = 0
                while i < len(self._key_details):
                    regex_for_rest = r"{}\S+\s+".format(regex_for_rest)
                    i += 1
                regex_for_rest = r"{}(?P<REST>\S.*\S|\S+)".format(regex_for_rest)
                re_for_rest = re.compile(regex_for_rest)
```

Reply:
done.